language/tools/resource-viewer/src/lib.rs

```diff
@@ -226,6 +226,69 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
     Ok(())
 }
 
+impl serde::Serialize for AnnotatedMoveStruct {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            let mut s = serializer.serialize_map(Some(self.value.len()))?;
+            for (f, v) in &self.value {
+                s.serialize_entry(f, v)?
+            }
+            s.end()
+        } else {
+            let mut s = serializer.serialize_tuple(self.value.len())?;
+            for (_f, v) in &self.value {
+                s.serialize_element(v)?
+            }
+            s.end()
+        }
+    }
+}
+
+impl serde::Serialize for AnnotatedMoveValue {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use AnnotatedMoveValue::*;
+        match self {
+            U8(n) => serializer.serialize_u8(*n),
+            U64(n) => serializer.serialize_u64(*n),
+            U128(n) => {
+                // TODO: we could use serializer.serialize_u128 here, but it requires the serde_json
+                // arbitrary_precision, which breaks some existing json-rpc test. figure
+                // out what's going on or come up with a better workaround
+                if let Ok(i) = u64::try_from(*n) {
+                    serializer.serialize_u64(i)
+                } else {
+                    serializer.serialize_bytes(&n.to_le_bytes())
+                }
+            }
+            Bool(b) => serializer.serialize_bool(*b),
+            Address(a) => a.serialize(serializer),
+            Vector(t, vals) => {
+                assert_ne!(t, &TypeTag::U8);
+                let mut vec = serializer.serialize_seq(Some(vals.len()))?;
+                for v in vals {
+                    vec.serialize_element(v)?;
+                }
+                vec.end()
+            }
+            Bytes(v) => {
+                // try to deserialize as utf8, fall back to hex with 0x prefix if we can't
```

Comment:
Are utf8 strings ever going to become first class citizens in move?

Reply:
Hacks like this make me think it's time to finally implement a string type :).

Reply:
There have been a number of times i've told folks to do string validation onchain so having some sort of string type would have made that easier/possible :)